lib/peg.d.ts

```diff
@@ -1185,6 +1185,12 @@ export interface BuildOptionsBase {
 }
 
 export interface ParserBuildOptions extends BuildOptionsBase {
+  /**
+   * Extensions may need to the caller to pass in otherwise-unknown options.
+   * ts-pegjs has an example in its README.md.
+   */
+  [extensionOpts: string]: any;
```

Comment:
Won't break much, but in general index signatures are [not very stable](https://github.com/microsoft/TypeScript/issues?q=is%3Aissue+is%3Aopen+index+signature). The correct approach to pass options to plugins looks like this:

```
peggy({
    // peggy options
    plugins: [
        goodPlugin({
            // plugin options
        })
    ],
})
```

A good real-world example is `webpack`'s API vs `vite`'s API.

Reply:
I agree that this would have been the right approach if we were designing the API, but I don't think breaking existing plugins is warranted for this change.

Reply:
Let's make a plan on how to phase out old API. A good start would be to review [this PR](https://github.com/polkovnikov-ph/peggy/pull/1). It highlights several problems of typing the existing API.

Reply:
I agree we need a plan, and I'm ok with moving to typescript as a part of a larger breaking change.  I think the next step is a discussion of how we generate code (which will cause a massive downstream breakage if we change anything), then figuring out what else we're going to break at the same time.  I'm working on writing something up about code generation.

Reply:
Linked PR adds TS without breaking changes (although I didn't test it). The point in reviewing the PR is to get an idea of where types get especially gnarly, and of which things we'd like to avoid.